agent/lib/kontena/launchers/etcd.rb

```diff
@@ -134,12 +143,83 @@ def create_container(image, info)
         }
       )
       container.start
-      Celluloid::Notifications.publish('dns:add', {id: container.id, ip: weave_ip, name: 'etcd.kontena.local'})
+      Celluloid::Notifications.publish('dns:add', {id: container.id, ip: weave_ip(info), name: 'etcd.kontena.local'})
       info "started etcd service"
       @running = true
       container
     end
 
+    # Removes possible previous member with the same IP
+    #
+    # @param [String] node weave ip
+    # @return [String] the state of the cluster member
+    def update_membership(info)
+      info "checking if etcd previous membership needs to be updated"
+      tries = info['grid']['initial_size']
+      node_number = info['node_number']
+      peer_url = "http://#{weave_ip(info)}:2380"
+      client_url = "http://#{weave_ip(info)}:2379"
+
+      begin
+        etcd_host = "http://10.81.0.#{tries}:2379/v2/members"
+
+        info "connecting to existing etcd at #{etcd_host}"
+        connection = Excon.new(etcd_host)
+        members = JSON.parse(connection.get().body)
+        peer_found = false
+        members['members'].each do |member|
+          if member['peerURLs'].include?(peer_url) && member['clientURLs'].include?(client_url)
+            # When there's both peer and client URLs, the given peer has been a member of the cluster
+            # and needs to be replaced
+            delete_membership(connection, member['id'])
+            sleep 1
+            add_membership(connection, peer_url)
+            sleep 1
+            return 'existing'
+          elsif member['peerURLs'].include?(peer_url) && !member['clientURLs'].include?(client_url)
+            peer_found = true
+          end
+        end
+
+        unless peer_found
+          info "previous member info not found at all, adding"
+          add_membership(connection, peer_url)
+        end
+        
+      rescue Excon::Errors::Error => exc
+        tries -= 1
+        if tries > 0
+          info "retrying next etcd host"
+          retry
+        else
+          error "cannot remove previous etcd membership info"
+          log_error exc
+        end
+      end
+      'new'
+    end
+
+
+    # Deletes membership of given etcd peer
+    #
+    # @param [Excon::Connection] etcd HTTP members API connection
+    # @param [String] id of the peer to be removed
+    def delete_membership(connection, id)
+      info "Removing existing etcd membership info with id #{id}"
+      connection.delete(:path => "/v2/members/#{id}")
+    end
+
+    ##
+    # Add new peer membership
+    #
+    # @param [Excon::Connection] etcd HTTP members API connection
+    # @param [String] The peer URL of the new peer to be added to the cluster
+    def add_membership(connection, peer_url)
+      info "Adding new etcd membership info with peer URL #{peer_url}"
+      connection.post(:body => JSON.generate({"peerURLs": [peer_url]}),
+                      :headers => { "Content-Type" => "application/json" })
```

Comment:
Minor style guide issue here. If string does not contain any variables use single quotation marks